File: governor/scoring/rubric.py
"""Scoring engine for Governor task reviews.

Implements a base-plus-excellence scoring model with configurable deductions.
"""

import json
import os
from typing import Any, Dict, List, Optional


class ScoringRubric:
    """Configurable scoring rubric engine.

    The default scoring model:
    - Base score: 85
    - Excellence: up to +15 (value add, requires evidence)
    - Deductions: uncapped negative points

    Usage::

        rubric = ScoringRubric()
        score = rubric.score(task_data, review_data)
    """

    def __init__(self, rubric_path: Optional[str] = None) -> None:
        if rubric_path is None:
            rubric_path = os.path.join(
                os.path.dirname(__file__), "rubrics", "default_rubric.json"
            )
        try:
            with open(rubric_path, "r", encoding="utf-8") as f:
                self._rubric = json.load(f)
        except FileNotFoundError:
            raise ValueError(
                f"Scoring rubric file not found: {rubric_path}. "
                "Pass a valid path or use the default rubric."
            )
        except json.JSONDecodeError as exc:
            raise ValueError(
                f"Scoring rubric is not valid JSON ({rubric_path}): {exc}"
            ) from exc
        self._categories = self._rubric.get("categories", {})
        self._category_max_points = {
            name: int(defn.get("max_points", 0))
            for name, defn in self._categories.items()
            if isinstance(defn, dict)
        }

    @property
    def base_score(self) -> int:
        return self._rubric.get("base_score", 85)

    @property
    def excellence_max(self) -> int:
        return self._rubric.get("excellence_max", 15)

    def score(
        self,
        categories: Dict[str, int],
        deductions: Optional[List[Dict[str, Any]]] = None,
        excellence: int = 0,
    ) -> Dict[str, Any]:
        """Calculate a final score.

        Args:
            categories: Dict of category name -> points awarded (0 to max).
            deductions: List of dicts with 'type' and 'points' keys.
            excellence: Points for excellence (0 to excellence_max).

        Returns:
            Dict with: base_score, categories, deductions, excellence,
            final_score, rating.
        """
        deductions = deductions or []

        self._validate_categories(categories)

        # Sum category scores, each capped at rubric-defined max points.
        category_total = 0
        for category_name, points in categories.items():
            max_points = self._category_max_points[category_name]
            category_total += min(int(points), max_points)
        category_total = min(category_total, self.base_score)

        # Sum deductions — clamp each to >= 0 so negative values cannot
        # inject positive points into the final score.
        deduction_total = sum(max(0, d.get("points", 0)) for d in deductions)

        # Cap excellence and apply evidence gate
        excellence = min(int(excellence), self.excellence_max)
        evidence_gate = self._rubric.get("evidence_gate", 80)
        if category_total < evidence_gate:
            excellence = 0  # Gate: base must meet threshold for excellence

        final_score = category_total + excellence - deduction_total
        final_score = max(0, min(final_score, self.base_score + self.excellence_max))

        return {
            "base_score": self.base_score,
            "category_total": category_total,
            "categories": categories,
            "deductions": deductions,
            "deduction_total": deduction_total,
            "excellence": excellence,
            "final_score": final_score,
            "rating": self._rating(final_score),
        }

    def _validate_categories(self, categories: Dict[str, int]) -> None:
        configured = set(self._category_max_points.keys())
        provided = set(categories.keys())
        unknown = sorted(provided - configured)
        if unknown:
            raise ValueError(
                f"Unknown scoring categories: {unknown}. "
                f"Allowed categories: {sorted(configured)}"
            )

        for category_name, points in categories.items():
            if not isinstance(points, (int, float)):
                raise ValueError(
                    f"Category '{category_name}' points must be numeric, got {type(points).__name__}"
                )
            if points < 0:
                raise ValueError(f"Category '{category_name}' points must be >= 0")

    def _rating(self, score: int) -> str:
        thresholds = self._rubric.get("rating_thresholds", {})
        if score >= thresholds.get("EXCEPTIONAL", 95):
            return "EXCEPTIONAL"
        elif score >= thresholds.get("EXCELLENT", 85):
            return "EXCELLENT"
        else:
            return "NEEDS_IMPROVEMENT"

File: governor-ts/src/scoring/rubric.ts
/**
 * Scoring engine for Governor task reviews.
 *
 * Implements a base-plus-excellence scoring model with configurable
 * deductions. Wire-compatible with the Python ScoringRubric.
 *
 * @example
 * ```ts
 * import { ScoringRubric } from "@governor/core";
 *
 * const rubric = new ScoringRubric();
 * const result = rubric.score(
 *   { completion_gate: 20, core_execution: 18, code_quality: 22, documentation_quality: 18 },
 *   [{ type: "linter_errors", points: 5 }],
 *   10,
 * );
 * console.log(result.final_score); // 83
 * ```
 */

import defaultRubric from "./rubrics/default_rubric.json" with { type: "json" };

// ------------------------------------------------------------------
// Types
// ------------------------------------------------------------------

export interface RubricCategoryDef {
  max_points: number;
  description?: string;
}

export interface RubricDef {
  rubric_name?: string;
  version?: string;
  description?: string;
  base_score: number;
  excellence_max: number;
  evidence_gate?: number;
  categories: Record<string, RubricCategoryDef>;
  rating_thresholds?: Record<string, number>;
  deduction_types?: Record<string, { points: number; description?: string }>;
}

export interface Deduction {
  type: string;
  points: number;
}

export interface ScoreResult {
  base_score: number;
  category_total: number;
  categories: Record<string, number>;
  deductions: Deduction[];
  deduction_total: number;
  excellence: number;
  final_score: number;
  rating: string;
}

// ------------------------------------------------------------------
// ScoringRubric
// ------------------------------------------------------------------

export class ScoringRubric {
  private readonly _rubric: RubricDef;
  private readonly _categoryMaxPoints: Record<string, number>;

  constructor(rubric?: RubricDef) {
    this._rubric = rubric ?? (defaultRubric as RubricDef);
    this._categoryMaxPoints = {};
    for (const [name, defn] of Object.entries(this._rubric.categories)) {
      if (defn && typeof defn === "object" && "max_points" in defn) {
        this._categoryMaxPoints[name] = defn.max_points;
      }
    }
  }

  get baseScore(): number {
    return this._rubric.base_score ?? 85;
  }

  get excellenceMax(): number {
    return this._rubric.excellence_max ?? 15;
  }

  /**
   * Calculate a final score.
   *
   * @param categories - Category name to points awarded (0 to max).
   * @param deductions - List of deduction objects with type and points.
   * @param excellence - Points for excellence (0 to excellence_max).
   */
  score(
    categories: Record<string, number>,
    deductions: Deduction[] = [],
    excellence: number = 0,
  ): ScoreResult {
    this._validateCategories(categories);

    // Sum category scores, each capped at rubric-defined max points.
    let categoryTotal = 0;
    for (const [name, points] of Object.entries(categories)) {
      const maxPts = this._categoryMaxPoints[name];
      categoryTotal += Math.min(points, maxPts);
    }
    categoryTotal = Math.min(categoryTotal, this.baseScore);

    // Sum deductions -- clamp each to >= 0.
    const deductionTotal = deductions.reduce(
      (sum, d) => sum + Math.max(0, d.points ?? 0),
      0,
    );

    // Cap excellence and apply evidence gate.
    let cappedExcellence = Math.min(Math.floor(excellence), this.excellenceMax);
    const evidenceGate = this._rubric.evidence_gate ?? 80;
    if (categoryTotal < evidenceGate) {
      cappedExcellence = 0;
    }

    const rawScore = categoryTotal + cappedExcellence - deductionTotal;
    const finalScore = Math.max(0, Math.min(rawScore, this.baseScore + this.excellenceMax));

    return {
      base_score: this.baseScore,
      category_total: categoryTotal,
      categories,
      deductions,
      deduction_total: deductionTotal,
      excellence: cappedExcellence,
      final_score: finalScore,
      rating: this._rating(finalScore),
    };
  }

  private _validateCategories(categories: Record<string, number>): void {
    const configured = new Set(Object.keys(this._categoryMaxPoints));
    const provided = new Set(Object.keys(categories));

    const unknown = [...provided].filter((k) => !configured.has(k)).sort();
    if (unknown.length > 0) {
      throw new Error(
        `Unknown scoring categories: [${unknown.join(", ")}]. ` +
          `Allowed categories: [${[...configured].sort().join(", ")}]`,
      );
    }

    for (const [name, points] of Object.entries(categories)) {
      if (typeof points !== "number") {
        throw new Error(
          `Category '${name}' points must be numeric, got ${typeof points}`,
        );
      }
      if (points < 0) {
        throw new Error(`Category '${name}' points must be >= 0`);
      }
    }
  }

  private _rating(score: number): string {
    const thresholds = this._rubric.rating_thresholds ?? {};
    if (score >= (thresholds.EXCEPTIONAL ?? 95)) return "EXCEPTIONAL";
    if (score >= (thresholds.EXCELLENT ?? 85)) return "EXCELLENT";
    return "NEEDS_IMPROVEMENT";
  }
}
